### ISSUE TITLE
[question] Revisions with different dependencies in requires

### ISSUE THREAD

**Issue by zawlin** [conan-io/conan#9912]
I have a question about how revisions are supposed to work with packages that can produce multiple binaries depending on different requires. For example, we need to build opencv with cuda 10 and 11. Without revisions, the workflow is

* Create conanfile.py with opencv with `requires = 'cuda/10.0'`
* buid -> export pkg 
* edit conanfile.py and update `requires = 'cuda/11.0'`
* build -> export pkg
* if I want to use opencv binary with cuda 11, from the consuming library, I just needs to specify `requires = ('opencv/4.5.3','cuda/11.0')`. Same for cuda 10.

However, with revisions enabled, this workflow no longer works because once I edited the conanfile.py, it fixes the revision. From the client, I can no longer have a simple `requires = ('opencv/4.5.3','cuda/11.0')`. I must now have `requires = ('opencv/4.5.3#RREF','cuda/11.0')`, if the latest revision is not cuda 11.0. What is the way to avoid having explicit RREF yet still be able to choose different cuda versions?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @zawlin 

> However, with revisions enabled, this workflow no longer works because once I edited the conanfile.py, it fixes the revision. From the client, I can no longer have a simple requires = ('opencv/4.5.3','cuda/11.0'). I must now have requires = ('opencv/4.5.3#RREF','cuda/11.0'), if the latest revision is not cuda 11.0.

What is the problem that you are seeing? Is it "missing binary" or is it a "version conflict"? 

When you are doing some change to the recipe, and changing it, it will have  a different revision. The different binaries that belong to that revision should be built. If the new revision needs to build still with both ``cuda/10`` and ``cuda/11``, you should build those 2 binaries. The binary that you had for the previous revision is no longer valid, because once there is a source change, anything can happen. The way you can build different flavors of ``opencv`` is with:

```bash
$ conan create . opencv/version@user/channel --require-override=cuda/11.0@
$ conan create . opencv/version@user/channel --require-override=cuda/10.0@
```

You might also want to define better in the recipe something like ``requires = "cuda/[>10 <12]"``, to represent that such is a valid range of cuda versions for this particular version and revision of the ``opencv`` recipe. That also allows to automatically use new cuda versions without modifying the recipe.

**Comment by zawlin** [comment]
The error I am seeing is "missing binary". Thanks for your suggestion, I understand now how it should work and tested it with my recipes.

However, I run into a bit of problem. I use the following to create the package and quickly test the scenario. My recipe does not yet work with "conan create"

1. Create zlib 1.2.8
    * conan install .
    * conan build .
    * conan export-pkg .
2. Create zlib 2.2.8
    * conan install .
    * conan build .
    * conan export-pkg .
3. Create png -> with zlib 2.2.8
    * conan install . --require-override=zlib/2.2.8
    * conan build .
    * conan export-pkg .
4. Create png -> with zlib 1.2.8
    * conan install . --require-override=zlib/1.2.8
    * conan build .
    * conan export-pkg . <---this step errors with
~~~
Version ranges solved
    Version range '1> <3' required by 'png/1.6.29' resolved to 'zlib/2.2.8' in local cache

Packaging to a304e0c2d77679b0c4bc3097621e0b71389f53f8
ERROR: Package already exists. Please use --force, -f to overwrite it
~~~

It's a bit strange what's happening at step 4 because it appears during the export, the override is not working correctly although it resolves correctly in "install" stage? It works in any of the following scenarios:
* I perform `conan remove "*"` and then conan install from remote
* I create the png with zlib 1.2.8 first and then create png with zlib 2.2.8.

**Comment by memsharded** [comment]
It seems you are using the same ``conan install . --require-override=zlib/1.2.8`` in both steps 3 and 4?

It is very possible that the ``export-pkg`` is missing the ``--require-override`` argument, as it was mainly designed for ``create`` and ``install``, but not for ``export-pkg``. In that case, we should add it

**Comment by zawlin** [comment]
It was a typo.  I was using the correct override in install. I think yes export-pkg is using the default greedy resolution which resolves to 2.2.8 and there is now no option to specify override.

**Comment by memsharded** [comment]
Ok, makes sense. I would recommend the ``conan create`` flow while this is being added. It should be equivalent to your install + build + export-pkg

**Comment by memsharded** [comment]
This will possible go together with https://github.com/conan-io/conan/pull/10046, we will investigate if using lockfiles is a better solution than ``--require-override`` arg, which might not be the best UI

**Comment by memsharded** [comment]
This will not fit in a 1.X scope, as https://github.com/conan-io/conan/pull/10046 has already been moved to 2.0-alpha, as its solution requires the new lockfiles. Moving it to 2.0 alpha stage

**Comment by memsharded** [comment]
This will finally be implemented using lockfiles. Added by https://github.com/conan-io/conan/pull/10713, will be released in next 2.0-alpha.6
